### PR TITLE
Fixed Wrong loss computation for VisionEncoderDecoderModel

### PR DESCRIPTION
**Pull request by Flakes342**
# What does this PR do?

Fixes #40111 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #40111 
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

This PR addresses a critical bug in `VisionEncoderDecoderModel` where loss computation was incorrect due to double-shifting of labels, causing poor training convergence since version 4.50.0.

## Problem

The issue occurred because:
1. `decoder_input_ids` is already shifted right w.r.t the labels in the model
2. The `ForCausalLMLoss` function also shifts labels left internally
3. This double-shifting resulted in misaligned logits and labels, leading to incorrect loss computation

## Solution

- **When `decoder_input_ids` is None**: Manually compute loss using properly shifted logits and labels to avoid double-shifting
- **When `decoder_input_ids` is provided**: Use existing loss function but add a warning about proper alignment
- This ensures correct loss computation in both scenarios while maintaining backward compatibility

### Core Changes
- Modified loss computation logic in `VisionEncoderDecoderModel.forward()`
- Added conditional path for manual loss calculation when `decoder_input_ids` is None
- Added warning when `decoder_input_ids` is manually provided

### Test Coverage
- `test_loss_calculation_without_decoder_input_ids`: Validates loss computation without decoder_input_ids
- `test_loss_calculation_paths_comparison`: Ensures both loss calculation paths produce valid results
- Tests added to all relevant encoder-decoder test classes

## Before/After Behavior

**Before**: Double-shifting caused misaligned logits/labels → poor training convergence
**After**: Proper alignment ensures correct loss computation → normal training convergence

## Testing

- All existing tests pass
- New tests validate both loss computation paths
- Backward compatibility maintained
- Warning system alerts users about manual decoder_input_ids usage

I am very new to open source contribution so please let me know if there are any issues or any fixing is required. Thanks!

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @qubvel

**Comment by Flakes342** [comment]
Hi @qubvel, fixed all the formatting errors best to my knowledge..awaiting your review. Thanks!

**Comment by qubvel** [comment]
Thanks for the fix! Let's remove `~tmp` folder from the PR

**Comment by Flakes342** [comment]
Sorry about the ~tmp folder..removed it from the PR

File: src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py [hunk 1]
```diff
@@ -572,12 +575,33 @@ def forward(
...
+
+                loss_fct = CrossEntropyLoss(ignore_index=-100)
+                loss = loss_fct(shift_logits.view(-1, self.decoder.config.vocab_size), shift_labels.view(-1))
+
+                loss = self.loss_function(
+                    logits=logits,
+                    labels=labels,
+                    vocab_size=self.decoder.config.vocab_size,
+                    num_items_in_batch=num_items_in_batch,
+                )
```
Comment (qubvel) [inline src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py]: should we compute loss twice?

File: src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py [hunk 2]
```diff
@@ -572,12 +575,33 @@ def forward(
...
-                logits=logits,
-                labels=labels,
-                vocab_size=self.decoder.config.vocab_size,
-                num_items_in_batch=num_items_in_batch,
-            )
+            if decoder_input_ids is None and original_labels_clone is not None:
+                shift_logits = logits[..., :-1, :].contiguous()
+                shift_labels = original_labels_clone[..., 1:].contiguous()
+
+                loss_fct = CrossEntropyLoss(ignore_index=-100)
```
Comment (qubvel) [inline src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py]:
```suggestion
                loss_function = CrossEntropyLoss(ignore_index=-100)
```

**Review by qubvel — COMMENTED** [review]
Thanks for updating, a few comments

File: src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py (hunk 1 above)
Comment (Flakes342) [reply src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py]: thanks for pointing that out, we should use only CrossEntropy loss in case the decoder_input_ids is none. Also, fixing as per the loss_function naming suggestion and resubmitting the PR

File: src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py [hunk 3]
```diff
@@ -572,12 +575,29 @@ def forward(
...
+                        "Typically it should be: shift_tokens_right(labels, pad_token_id, decoder_start_token_id). "
+                        "If misaligned, this may cause poor training convergence."
+                    )
+                # Using existing loss function
+                loss = self.loss_function(
+                    logits=logits,
+                    labels=labels,
+                    vocab_size=self.decoder.config.vocab_size,
+                    num_items_in_batch=num_items_in_batch,
+                )
```
Comment (qubvel) [inline src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py]: btw can we simplify? or it's not equivalent

```suggestion
            if decoder_input_ids is None and labels is not None:
                logits = logits[..., :-1, :].contiguous()
                labels = labels.clone()[..., 1:].contiguous()

            if decoder_input_ids is not None:
                logger.warning(
                    "decoder_input_ids was provided manually. For training, ensure that "
                    "decoder_input_ids is correctly shifted with respect to labels. "
                    "Typically it should be: shift_tokens_right(labels, pad_token_id, decoder_start_token_id). "
                    "If misaligned, this may cause poor training convergence."
                )

            # Using existing loss function
            loss = self.loss_function(
                logits=logits,
                labels=labels,
                vocab_size=self.decoder.config.vocab_size,
                num_items_in_batch=num_items_in_batch,
            )
```

Comment (Flakes342) [reply src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py]: > btw can we simplify? or it's not equivalent

Uh the core issue was that 'self.loss_function()' internally does its own shifting logic, which causes double shifting when combined with manually shifted inputs when decoder_input_ids is None. The suggested approach would still pass shifted logits to self.loss_function(), which can cause the same issue again I think

**Comment by Flakes342** [comment]
what could be this issue @qubvel?

File: src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py (hunk 3 above)
Comment (qubvel) [reply src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py]: ok, got it

**Comment by qubvel** [comment]
It was an unrelated failure. I merged main to make sure tests pass

File: src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py (hunk 3 above)
Comment (qubvel) [reply src/transformers/models/vision_encoder_decoder/modeling_vision_encoder_decoder.py]: there is a `shift_labels` attribute for the `self.loss_function`, if you provide `labels` that way they will not be shifted

e.g.

```py
if ...
            # not shifted inside loss
            loss = self.loss_function(
                logits=logits,
                shift_labels=labels,  # <-------------------------- diff
                vocab_size=self.decoder.config.vocab_size,
                num_items_in_batch=num_items_in_batch,
            )
else:
            loss = self.loss_function(
                logits=logits,
                labels=labels,
                vocab_size=self.decoder.config.vocab_size,
                num_items_in_batch=num_items_in_batch,
            )
```

**Comment by alephpi** [comment]
Hi, when would this PR be merged? I'm currently troubled by this bug.

**Comment by Flakes342** [comment]
> Hi, when would this PR be merged? I'm currently troubled by this bug.

Please have a look at the shift_labels attribute of the loss function referred above by @qubvel
